File: scripts/replace_old_instances.rb
require 'aws-sdk'
require 'awesome_print'

unless ARGV[0]
  puts 'Specify auto scaling group name.'
  exit 1
end
ASG_NAME = ARGV[0]

as = Aws::AutoScaling::Client.new
asg = Aws::AutoScaling::AutoScalingGroup.new(ASG_NAME, client: as)

if asg.desired_capacity * 2 > asg.max_size
  puts "Can not scale out to replace instances in #{ASG_NAME}."
  puts 'Consider increasing the max size of it.'
  exit 1
end

asg.set_desired_capacity(
  desired_capacity: asg.desired_capacity * 2,
  honor_cooldown: true,
)

# scaling policy will decrease desired capacity soon.

File: scripts/replace_old_instances.js
const AWS = require('aws-sdk');
const ArgumentParser = require('argparse').ArgumentParser;

AWS.config.region = process.env.AWS_DEFAULT_REGION;

const AutoScaling = new AWS.AutoScaling();

function parseArguments() {
  const parser = new ArgumentParser({
    description: 'This script scale out specified auto scaling group to flip out old instances.',
  });
  parser.addArgument(['-n', '--name'], {
    help: 'name of a auto scaling group',
    required: true,
  });
  return parser.parseArgs();
}

function validateAsgCapacity(asg) {
  if (!asg) {
    console.error('This autoscaling group does not exist...');
    process.exit(1);
  }
  if (asg.DesiredCapacity * 2 > asg.MaxSize) {
    console.error(`Can not scale out to replace instances. Consider increasing the max capacity.`);
    process.exit(1);
  }
  console.log(`Current desired capacity is ${asg.DesiredCapacity}`);

  return asg;
}

function multipleDesiredCapacity(asg) {
  console.log(`Scale out to ${asg.DesiredCapacity * 2}.`);

  return AutoScaling.setDesiredCapacity({
    AutoScalingGroupName: asg.AutoScalingGroupName,
    DesiredCapacity: asg.DesiredCapacity * 2,
    HonorCooldown: true,
  }).promise();
}

const asgName = parseArguments().name;

AutoScaling.describeAutoScalingGroups({
  AutoScalingGroupNames: [asgName],
}).promise()
  .then(data => data.AutoScalingGroups[0])
  .then(validateAsgCapacity)
  .then(multipleDesiredCapacity);
